File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if root==None:
            return 0
        maxval=root.val
        return 1+self.check(root.left,maxval)+self.check(root.right,maxval)
    def check(self,root:TreeNode,maxval:int)->int:
        if root==None:
            return 0
        res=0
        if root.val>=maxval:
            res+=1
        maxval=max(maxval,root.val)
        res+=self.check(root.left,maxval)+self.check(root.right,maxval)
        return res



        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> left;
        for(int i=0;i<nums.size();i++)
        {
            if(i==0)
            {
                left.push_back(1);
                continue;
            }
            int res=left[i-1]*nums[i-1];
            left.push_back(res);
        }
        
        vector<int> right;
        for(int i=nums.size()-1;i>=0;i--)
        {
            if(i==nums.size()-1)
            {
                right.push_back(1);
                continue;
            }
            int res=nums[i+1]*right[right.size()-1];
            right.push_back(res);
        }
        reverse(right.begin(),right.end());
        // for(int i=0;i<right.size();i++)
        // {
        //     cout<<right[i];
        // }
        vector<int> res;
        for(int i=0;i<nums.size();i++)
        {
            res.push_back(left[i]*right[i]);
        }
        return res;

    }
};
